src/perf/lib/ThroughputClient.cpp

```diff
@@ -0,0 +1,369 @@
+#include "quic_trace.h"
+
+ThroughputClient::ThroughputClient(
+    ) {
+    QuicZeroMemory(&LocalIpAddr, sizeof(LocalIpAddr));
+    if (Session.IsValid()) {
+        Session.SetAutoCleanup();
+    }
+}
+
+QUIC_STATUS
+ThroughputClient::Init(
+    _In_ int argc,
+    _In_reads_(argc) _Null_terminated_ char* argv[]
+    ) {
+    if (!Session.IsValid()) {
+        return Session.GetInitStatus();
+    }
+
+    Port = THROUGHPUT_DEFAULT_PORT;
+    TryGetValue(argc, argv, "port", &Port);
+
+    TryGetValue(argc, argv, "encrypt", &UseEncryption);
+
+    const char* Target;
+    if (!TryGetValue(argc, argv, "target", &Target)) {
+        WriteOutput("Must specify '-target' argument!\n");
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+
+    uint16_t Ip;
+    if (TryGetValue(argc, argv, "ip", &Ip)) {
+        switch (Ip) {
+        case 4: RemoteFamily = AF_INET; break;
+        case 6: RemoteFamily = AF_INET6; break;
+        }
+    }
+
+    TryGetValue(argc, argv, "length", &Length);
+
+    const char* LocalAddress = nullptr;
+    if (TryGetValue(argc, argv, "bind", &LocalAddress)) {
+        if (!ConvertArgToAddress(LocalAddress, 0, &LocalIpAddr)) {
+            WriteOutput("Failed to decode IP address: '%s'!\nMust be *, a IPv4 or a IPv6 address.\n", LocalAddress);
+            return QUIC_STATUS_INVALID_PARAMETER;
+        }
+    }
+
+    // TODO: Core, since we need to support kernel mode
+#ifdef QUIC_COMPARTMENT_ID
+    uint16_t CompartmentId;
+    if (TryGetValue(argc, argv, "comp",  &CompartmentId)) {
+        NETIO_STATUS status;
+        if (!NETIO_SUCCESS(status = QuicCompartmentIdSetCurrent(CompartmentId))) {
+            WriteOutput("Failed to set compartment ID = %d: 0x%x\n", CompartmentId, status);
+            return QUIC_STATUS_INVALID_PARAMETER;
+        } else {
+            WriteOutput("Running in Compartment %d\n", CompartmentId);
+        }
+    }
+#endif
+
+#ifdef QuicSetCurrentThreadAffinityMask
+    uint8_t CpuCore;
+    if (TryGetValue(argc, argv, "core",  &CpuCore)) {
+        QuicSetCurrentThreadAffinityMask((DWORD_PTR)(1ull << CpuCore));
+    }
+#endif
+
+    TryGetValue(argc, argv, "sendbuf", &UseSendBuffer);
+
+    IoSize = UseSendBuffer ? THROUGHPUT_DEFAULT_IO_SIZE_BUFFERED : THROUGHPUT_DEFAULT_IO_SIZE_NONBUFFERED;
+    TryGetValue(argc, argv, "iosize", &IoSize);
+
+    IoCount = UseSendBuffer ? THROGHTPUT_DEFAULT_SEND_COUNT_BUFFERED : THROUGHPUT_DEFAULT_SEND_COUNT_NONBUFFERED;
+    TryGetValue(argc, argv, "iocount", &IoCount);
+
+    size_t Len = strlen(Target);
+    TargetData.reset(new char[Len + 1]);
+    QuicCopyMemory(TargetData.get(), Target, Len);
+    TargetData[Len] = '\0';
+
+    BufferAllocator.Initialize(IoSize);
+
+    return QUIC_STATUS_SUCCESS;
+}
+
+struct ShutdownWrapper {
+    HQUIC ConnHandle {nullptr};
+    ~ShutdownWrapper() {
+        if (ConnHandle) {
+            MsQuic->ConnectionShutdown(ConnHandle, QUIC_CONNECTION_SHUTDOWN_FLAG_NONE, 0);
+        }
+    }
+};
```

Comment:
I feel like this doesn't need to be in this file, and possibly, instead, in msquic.hpp.

Reply:
This isn't general purpose though. Its specific to how I'm initializing things. Sometimes you might need other types of handles with an RAII wrapper.